docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+
+3. “Something” needs to watch the cluster to see how the resource is supposed to be used and make that happen. Minions fall down here, it's more that the apiserver reaches out to them, but I think that’s an ok exception.  The model still mostly fits for GCE PDs. The kubelet watches the pod description and from the volume info in the pod definition it mounts and actually uses the storage. I think a different, but very similar, pattern should be generalized and should work well for externally reachable functionality (load balancers, external IP, etc) and other storage mechanisms.
+
+I'll work through a couple of examples of what I'm thinking. Say an external load balancer (conceptually should map to external IPs or whatever else) and an NFS server (should map to other storage, including GCE PDs)
+
+**External Load Balancer Example**
+
+If I install an external load balancer into my datacenter with some form of reachability into the cluster (aka it can reach pods via pod IP addresses) I believe that the cloud provider interface needs to tell the cluster about the load balancer via some sort of “name” or “label” or whatever detail we agree upon. That's step 1. Step 2 is, in my mind, creating a new resource in kube. This resource should be a mapping between a service/pod/label and the 'name'/'label' of the external device. Yes, I’m talking about a first class resource like a pod, service, minion, etc. Step 3 is that the external load balancer (or a plugin in front of the load balancer) should watch the api server for new assignments and make the magic configuration changes for my load balancer.
+
+**NFS Server Storage Example**
+
+As an example lets say I want to use an NFS server to expose 10 exports for pod storage. I believe that either an admin or the infrastructure itself should make all of the information needed to mount those 10 exports available to the API server. I think this should be done as a new ‘volume’ resource (yes, very like the volume portion of the pod definition).  So (for NFS) the cloud provider (or maybe a kube admin) would make available the servername, export name, and any credentials necessary to use the mount. Now the kube admin/operator/user should be able to define a pod which references the storage by name. He should NOT have to know all of the storage details (as the volume requires today). The apiserver should take the volumeMount from the pod definition, the volume resource, and make all of that available to the kubelet. The dev who wrote the pod was told to use 'eparis.volume.42' and that's all they know. The kublet however gets nfs-server.example.com:/export/42/eparis.volume.stage/ along with a kerberos keytab from the apiserver. It can now mount and use the NFS mount.  I believe the concept should be able to map pretty cleanly to GCE PDs, or any storage concept (durable/empty/host are a little special, but not terribly outside the model.)
```

Comment:
Yes, one could think of that as "volume drivers", or auto-mount. I think ClusterHQ is promoting something along those lines. Filesystems are one area that probably developed all the abstractions we'll need long ago.